src/core/AxpyInterface.cpp

```diff
@@ -320,423 +316,504 @@ AxpyInterface<T>::AxpyInterface( AxpyType type, const DistMatrix<T>& X )
-    }
+      {
+	attachedForGlobalToLocal_ = true;
+	globalToLocalMat_ = &Z;
+      }
+    const Int p = Z.Grid ().Size ();
+
+#if MPI_VERSION>=3 && defined(EL_USE_IBARRIER_FOR_AXPY)
+#else
+    sentEomTo_.resize (p, false);
+    haveEomFrom_.resize (p, false);
+    eomSendRequests_.resize (p);
+#endif    
+    
+    // request objects
+    sendingRequest_.resize (p);
+    sendingData_.resize (p);
+    sendingReply_.resize (p);
+
+    // ready-to-send 
+    requestSendRequests_.resize (p);
+    replySendRequests_.resize (p);
+    dataSendRequests_.resize (p);
+      
+    dataVectors_.resize (p);  
+    requestVectors_.resize (p);  
+    replyVectors_.resize (p);
+  }
+
+  template < typename T >
+    void AxpyInterface < T >::Attach (AxpyType type,
+				      const DistMatrix < T > &X)
+  {
+    DEBUG_ONLY (CallStackEntry cse ("AxpyInterface::Attach"))
+      if (attachedForLocalToGlobal_ || attachedForGlobalToLocal_)
+      LogicError ("Must detach before reattaching.");
+
+    if (type == LOCAL_TO_GLOBAL)
+      {
+	LogicError ("Cannot update a constant matrix");
+      }
     else
-    {
-        attachedForGlobalToLocal_ = true;
-        globalToLocalMat_ = &X;
-    }
-
-    const Int p = X.Grid().Size();
-    sentEomTo_.resize( p, false );
-    haveEomFrom_.resize( p, false );
-
-    sendingData_.resize( p );
-    sendingRequest_.resize( p );
-    sendingReply_.resize( p );
-
-    dataVectors_.resize( p );
-    requestVectors_.resize( p );
-    replyVectors_.resize( p );
-
-    dataSendRequests_.resize( p );
-    requestSendRequests_.resize( p );
-    replySendRequests_.resize( p );
-
-    eomSendRequests_.resize( p );
-}
-
-template<typename T>
-void AxpyInterface<T>::Axpy( T alpha, Matrix<T>& Z, Int i, Int j )
-{
-    DEBUG_ONLY(CallStackEntry cse("AxpyInterface::Axpy"))
-    if( attachedForLocalToGlobal_ )
-        AxpyLocalToGlobal( alpha, Z, i, j );
-    else if( attachedForGlobalToLocal_ )
-        AxpyGlobalToLocal( alpha, Z, i, j );
+      {
+	attachedForGlobalToLocal_ = true;
+	globalToLocalMat_ = &X;
+      }
+    const Int p = X.Grid ().Size ();
+
+#if MPI_VERSION>=3 && defined(EL_USE_IBARRIER_FOR_AXPY)
+#else 
+    // eom   
+    sentEomTo_.resize (p, false);
+    haveEomFrom_.resize (p, false);
+    eomSendRequests_.resize (p);
+#endif
+
+    // ready-to-send 
+    sendingRequest_.resize (p);
+    sendingData_.resize (p);
+    sendingReply_.resize (p);
+
+    // ready-to-send 
+    requestSendRequests_.resize (p);
+    replySendRequests_.resize (p);
+    dataSendRequests_.resize (p);
+
+    dataVectors_.resize (p);
+    replyVectors_.resize (p);
+    requestVectors_.resize (p);
+   }
+
+  template < typename T >
+    void AxpyInterface < T >::Axpy (T alpha, Matrix < T > &Z, Int i, Int j)
+  {
+    DEBUG_ONLY (CallStackEntry cse ("AxpyInterface::Axpy"))
+      if (attachedForLocalToGlobal_)
+      AxpyLocalToGlobal (alpha, Z, i, j);
+    else if (attachedForGlobalToLocal_)
+      AxpyGlobalToLocal (alpha, Z, i, j);
```

Comment:
More indentation weirdness. It seems to be on purpose to indent after 'template' statements; I would recommend making the 'void' left aligned above.